mirgecom/initializers.py

```diff
@@ -462,6 +453,223 @@ def exact_rhs(self, discr, q, t=0.0):
+
+        self._nspecies = nspecies
+        self._dim = numdim
+        self._velocity = velocity
+        self._center = center
+        self._p0 = p0
+        self._rho0 = rho0
+        self._rhoamp = rhoamp
+        self._spec_y0s = spec_y0s
+        self._spec_centers = spec_centers
+        self._spec_amplitudes = spec_amplitudes
+
+    def __call__(self, t, x_vec, eos=IdealSingleGas()):
+        """
+        Create the lump-of-mass solution at time *t* and locations *x_vec*.
+
+        Note that *t* is used to advect the mass lump under the assumption of
+        constant, and uniform velocity.
+
+        Parameters
+        ----------
+        t: float
+            Current time at which the solution is desired
+        x_vec: numpy.ndarray
+            Nodal coordinates
+        eos: :class:`mirgecom.eos.GasEOS`
+            Equation of state class to be used in construction of soln (if needed)
+        """
+        print(f"len(x_vec) = {len(x_vec)}")
+        print(f"self._dim = {self._dim}")
+        if len(x_vec) != self._dim:
+            raise ValueError(f"Expected {self._dim}-dimensional inputs.")
+
+        amplitude = self._rhoamp
+        actx = x_vec[0].array_context
+
+        # mass0 = self._rho0
+        loc_update = t * self._velocity
+        lump_loc = self._center + loc_update
+        # coordinates relative to lump center
+        rel_center = make_obj_array(
+            [x_vec[i] - lump_loc[i] for i in range(self._dim)]
+        )
+        r = actx.np.sqrt(np.dot(rel_center, rel_center))
+        expterm = amplitude * actx.np.exp(- r ** 2)
+        mass = expterm + self._rho0
+        mom = self._velocity * make_obj_array([mass])
+        gamma = eos.gamma()
+        energy = (self._p0 / (gamma - 1.0)) + np.dot(mom, mom) / (2.0 * mass)
+
+        # process the species components independently
+        mass_fracs = None
+        if self._nspecies > 0:
+            lump_locs = make_obj_array([self._spec_centers[i] + loc_update
+                                       for i in range(self._nspecies)])
+            rel_poss = make_obj_array([x_vec - lump_locs[i]
+                                       for i in range(self._nspecies)])
+            r2s = make_obj_array([np.dot(rel_poss[i], rel_poss[i])
+                                  for i in range(self._nspecies)])
+            expterms = make_obj_array(
+                [self._spec_amplitudes[i] * actx.np.exp(- r2s[i])
+                 for i in range(self._nspecies)]
+            )
+            mass_fracs = make_obj_array([self._spec_y0s[i] + expterms[i]
+                                        for i in range(self._nspecies)])
+
+        return flat_obj_array(mass, energy, mom, mass_fracs)
+
+    def exact_rhs(self, discr, q, t=0.0):
+        """
+        Create the RHS for multlump-of-mass solution at time *t*, locations *x_vec*.
+
+        Note that this routine is only useful for testing under the condition of
+        uniform, and constant velocity field.
```

Comment:
Is it possible to create anything else using this IC?

Reply:
IMO it is wrong to refer to these as IC.  They can be *used* as IC, but in general they are just solution initializers that can be used at any time (e.g. for boundary solution at arbitrary time, or introducing an acoustic pulse). 

Regardless, the note is still valid because it reminds the user (me, in this case) - that the routine cannot be used to get the expected RHS for unsteady configurations.

Reply:
> Is it possible to create anything else using this IC?

Specifically, no.  If the user initializes Q using *only* this initializer, which will initialize Q with a uniform flow with disparate, uncoupled bumps in the scalar components, then this routine can be used to pick up the expected RHS at any time t >= t_initial, because it just does Q(r,t) = Q(r - Vt), where V is the uniform and constant velocity that was specified by the user at init time.

So if anything goes into the soln fields Q introducing or inducing an unsteadiness, then the assumption of constant, uniform V is wrong. This note is only here as a reminder to anyone who would use this function in a test (its only real use) to be cautious not to run afoul of the assumptions.

Clearly, if it is has raised a question for you, then the comment isn't clear enough?

Reply:
https://github.com/illinois-ceesd/mirgecom/pull/95/commits/ec6e5443ba60de55351a4360f1671e64390c4257#r562656000